{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Template\n",
    "\n",
    "This notebook allows users to find common and specific genes in their experiment of interest using an *existing* VAE model and selecting a template experiment that is included in the training compendium.\n",
    "\n",
    "This notebook will generate a `generic_gene_summary_<experiment id>.tsv` file that contains z-scores per gene that indicates how specific a gene is the experiment in question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/sophie/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from ponyo import utils, simulate_expression_data\n",
    "from sophie import (\n",
    "    process,\n",
    "    stats,\n",
    "    ranking,\n",
    ")\n",
    "# Remove later\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inputs\n",
    "\n",
    "User needs to fill in the [config file](config_new_experiment.tsv) following the instructions provided in the [readme]](README.md)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = \"config_example.tsv\"\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params[\"local_dir\"]\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# Metadata file that maps experiment ids to sample ids\n",
    "experiment_to_sample_filename = params[\"experiment_to_sample_filename\"]\n",
    "\n",
    "# Delimiter used in metadata file\n",
    "metadata_delimiter = params[\"metadata_delimiter\"]\n",
    "\n",
    "# Column header corresponding to experiment id and sample id\n",
    "experiment_id_colname = params[\"experiment_id_colname\"]\n",
    "sample_id_colname = params[\"sample_id_colname\"]\n",
    "\n",
    "# ID for template experiment to be selected\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params[\"num_simulated\"]\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params[\"vae_model_dir\"]\n",
    "\n",
    "# Size of the latent dimension\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "\n",
    "# Which DE method to use\n",
    "# We recommend that if data is RNA-seq then use DESeq2\n",
    "# If data is microarray then use Limma\n",
    "de_method = params[\"DE_method\"]\n",
    "\n",
    "# If using DE-seq, setting this parameter will\n",
    "# remove genes below a certain threshold\n",
    "count_threshold = params[\"count_threshold\"]\n",
    "\n",
    "# Metadata file that specifies which samples to keep for DE analysis (Optional)\n",
    "template_process_samples_filename = params[\"template_process_samples_filename\"]\n",
    "    \n",
    "# Metadata file that specifies sample grouping for DE analysis\n",
    "template_DE_grouping_filename = params[\"template_DE_grouping_filename\"]\n",
    "\n",
    "# Statistic to use to rank genes or pathways by\n",
    "# Choices are {} FILL IN\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files generated by this notebook\n",
    "\n",
    "# File storing normalized template experiment\n",
    "normalized_template_filename = params[\"normalized_template_filename\"]\n",
    "\n",
    "# File storing un-normalized template experiment for us in downstream DE analysis\n",
    "raw_template_filename = params[\"raw_template_filename\"]\n",
    "\n",
    "# File storing processed template experiment,\n",
    "# after samples have been selected for comparison in DE analysis\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Output summary file\n",
    "output_filename = params[\"output_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get unnormalized template experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "process.fetch_template_experiment(\n",
    "    normalized_compendium_filename,\n",
    "    experiment_to_sample_filename,\n",
    "    metadata_delimiter,\n",
    "    experiment_id_colname,\n",
    "    project_id,\n",
    "    sample_id_colname,\n",
    "    scaler_filename,\n",
    "    raw_template_filename,\n",
    "    normalized_template_filename\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/sophie/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/sophie/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments UPDATE COMMENT\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:\n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1]\n",
    "\n",
    "# REMOVE LATER\n",
    "NN_architecture = \"NN_test\"\n",
    "dataset_name = \"pre_model_seen_template\"\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))\n",
    "\n",
    "# Update simulated dir\n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "\n",
    "#------------\n",
    "simulate_expression_data.shift_template_experiment(\n",
    "    normalized_compendium_filename,\n",
    "    NN_architecture,\n",
    "    latent_dim,\n",
    "    dataset_name,\n",
    "    scaler,\n",
    "    experiment_to_sample_filename,\n",
    "    metadata_delimiter,\n",
    "    experiment_id_colname,\n",
    "    sample_id_colname,\n",
    "    project_id,\n",
    "    local_dir,\n",
    "    base_dir,\n",
    "    num_runs,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process template and simulated experiments\n",
    "\n",
    "* Remove samples not required for comparison\n",
    "* Make sure ordering of samples matches metadata for proper comparison\n",
    "* Make sure values are cast as integers if using DESeq\n",
    "* Filter lowly expressed genes if using DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "# Update simulated dir\n",
    "if not os.path.exists(template_process_samples_filename):\n",
    "    template_process_samples_filename = None\n",
    "\n",
    "if de_method == \"deseq\":\n",
    "    # Process template data\n",
    "    stats.process_samples_for_DESeq(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    # Process simulated data\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            template_process_samples_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            None,\n",
    "            template_process_samples_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pass simulated dir to R scripts below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: DelayedArray\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: BiocParallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in !(all.equal(colnames(expression_data), metadata$sample)) : \n",
      "  invalid argument type\n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: There were 12 warnings (use warnings() to see them)\n",
      "R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error in !(all.equal(colnames(expression_data), metadata$sample)) : \n",
      "  invalid argument type\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/sophie/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in !(all.equal(colnames(expression_data), metadata$sample)) : \n",
      "  invalid argument type\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error in !(all.equal(colnames(expression_data), metadata$sample)) : \n",
      "  invalid argument type\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i base_dir -i local_dir -i num_runs -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/sophie/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (de_method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank genes\n",
    "\n",
    "Genes are ranked by their \"generic-ness\" - how frequently these genes are changed across the simulated experiments using user-specific test statistic provided in the `col_to_rank_genes` params (i.e. log2 fold change)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/sophie/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "# Added\n",
    "if de_method == \"deseq\":\n",
    "    logFC_name = \"log2FoldChange\"\n",
    "    pvalue_name = \"padj\"\n",
    "else:\n",
    "    logFC_name = \"logFC\"\n",
    "    pvalue_name = \"adj.P.Val\"\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary table\n",
    "\n",
    "* Gene ID: Gene identifier (hgnc symbols for human data or PA number for *P. aeruginosa* data)\n",
    "* (Real): Statistics for template experiment\n",
    "* (Simulated): Statistics across simulated experiments\n",
    "* Number of experiments: Number of simulated experiments\n",
    "* Z-score: High z-score indicates that gene is more changed in template compared to the null set of simulated experiments (high z-score = highly specific to template experiment)\n",
    "* Percentile (simulated): percentile rank of the median(abs(log fold change)). So its the median absolute change for that gene across the 25 simulated experiments that is then converted to a percentile rank from 0 - 100. Where a higher percentile indicates that the gene was highly changed frequently and would suggest that the gene is more commonly DE.\n",
    "* Percent DE (simulated): the fraction of the simulated experiments in which that gene was found to be DE using (log fold change > 1 and adjusted p-value < 0.05). _Note:_ you may find that many genes have a 0 fraction. This is because there is some compression that happens when pushing data through the VAE so the variance of the simulated experiments is lower compared to the real experiment. We are aware of this limitation in the VAE and are looking at how to improve the variance and biological signal captured by the VAE, however we were still able to demonstrate that for now the VAE is able to simulate realistic looking biological experiments in our previous [paper](https://academic.oup.com/gigascience/article/9/11/giaa117/5952607).\n",
    "\n",
    "\n",
    "**Note:**\n",
    "* If using DESeq, genes with NaN in only the `Adj P-value (Real)` column are those genes flagged because of the `cooksCutoff` parameter. The cook's distance as a diagnostic to tell if a single sample has a count which has a disproportionate impact on the log fold change and p-values. These genes are flagged with an NA in the pvalue and padj columns of the result table.\n",
    "\n",
    "* If using DESeq with count threshold, some genes may not be present in all simulated experiments (i.e. the `Number of experiments (simulated)` will not equal the number of simulated experiments you specified in the beginning. This pre-filtering will lead to some genes found in few simulated experiments and so the background/null set for that gene is not robust. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "* If using DESeq without count threshold, some genes may still not be present in all simulated experiments (i.e. the `Number of experiments (simulated)`  will not equal the number of simulated experiments you specified in the beginning. If the gene is 0 expressed across all samples and thus automatically given an NA in `log fold change, adjusted p-value` columns. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "For more information you can read [DESeq FAQs](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VEZT</th>\n",
       "      <td>VEZT</td>\n",
       "      <td>6.561899e-04</td>\n",
       "      <td>9622.0</td>\n",
       "      <td>0.564027</td>\n",
       "      <td>0.564027</td>\n",
       "      <td>0.045076</td>\n",
       "      <td>4234.0</td>\n",
       "      <td>23.842514</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.236518</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>2</td>\n",
       "      <td>6301.055203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THAP9</th>\n",
       "      <td>THAP9</td>\n",
       "      <td>1.411167e-01</td>\n",
       "      <td>7741.0</td>\n",
       "      <td>0.399319</td>\n",
       "      <td>0.399319</td>\n",
       "      <td>0.784561</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.597049</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027406</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>2</td>\n",
       "      <td>6257.662855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC1</th>\n",
       "      <td>TTC1</td>\n",
       "      <td>2.812214e-02</td>\n",
       "      <td>8193.0</td>\n",
       "      <td>0.434710</td>\n",
       "      <td>0.434710</td>\n",
       "      <td>0.070044</td>\n",
       "      <td>1441.0</td>\n",
       "      <td>8.110848</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.115925</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>2</td>\n",
       "      <td>2698.733956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01730</th>\n",
       "      <td>LINC01730</td>\n",
       "      <td>1.301770e-02</td>\n",
       "      <td>13560.0</td>\n",
       "      <td>1.484307</td>\n",
       "      <td>1.484307</td>\n",
       "      <td>0.757420</td>\n",
       "      <td>303.0</td>\n",
       "      <td>1.701025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.050050</td>\n",
       "      <td>0.000591</td>\n",
       "      <td>2</td>\n",
       "      <td>2428.170510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIT</th>\n",
       "      <td>CIT</td>\n",
       "      <td>5.156598e-08</td>\n",
       "      <td>14996.0</td>\n",
       "      <td>2.899800</td>\n",
       "      <td>2.899800</td>\n",
       "      <td>0.254798</td>\n",
       "      <td>2050.0</td>\n",
       "      <td>11.541061</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142667</td>\n",
       "      <td>0.001702</td>\n",
       "      <td>2</td>\n",
       "      <td>1619.890114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CASR</th>\n",
       "      <td>CASR</td>\n",
       "      <td>1.926148e-01</td>\n",
       "      <td>15482.0</td>\n",
       "      <td>4.649033</td>\n",
       "      <td>4.649033</td>\n",
       "      <td>0.402032</td>\n",
       "      <td>10815.0</td>\n",
       "      <td>60.910217</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.903815</td>\n",
       "      <td>0.002424</td>\n",
       "      <td>2</td>\n",
       "      <td>1544.822271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CWC15</th>\n",
       "      <td>CWC15</td>\n",
       "      <td>6.812821e-01</td>\n",
       "      <td>11302.0</td>\n",
       "      <td>0.788082</td>\n",
       "      <td>0.788082</td>\n",
       "      <td>0.003572</td>\n",
       "      <td>3123.0</td>\n",
       "      <td>17.584770</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.187333</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>2</td>\n",
       "      <td>1474.512916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM110B</th>\n",
       "      <td>FAM110B</td>\n",
       "      <td>1.186656e-26</td>\n",
       "      <td>16007.0</td>\n",
       "      <td>21.693587</td>\n",
       "      <td>-21.693587</td>\n",
       "      <td>0.081285</td>\n",
       "      <td>9364.0</td>\n",
       "      <td>52.737411</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.614677</td>\n",
       "      <td>0.015747</td>\n",
       "      <td>2</td>\n",
       "      <td>1338.586384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DYNC1LI2</th>\n",
       "      <td>DYNC1LI2</td>\n",
       "      <td>4.908478e-02</td>\n",
       "      <td>5687.0</td>\n",
       "      <td>0.267402</td>\n",
       "      <td>0.267402</td>\n",
       "      <td>0.328713</td>\n",
       "      <td>882.0</td>\n",
       "      <td>4.962262</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088109</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>2</td>\n",
       "      <td>1324.782792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUL1</th>\n",
       "      <td>MUL1</td>\n",
       "      <td>6.487886e-01</td>\n",
       "      <td>1893.0</td>\n",
       "      <td>0.069030</td>\n",
       "      <td>-0.069030</td>\n",
       "      <td>0.755263</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.309789</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018416</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>2</td>\n",
       "      <td>1266.050242</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "VEZT            VEZT        6.561899e-04       9622.0   \n",
       "THAP9          THAP9        1.411167e-01       7741.0   \n",
       "TTC1            TTC1        2.812214e-02       8193.0   \n",
       "LINC01730  LINC01730        1.301770e-02      13560.0   \n",
       "CIT              CIT        5.156598e-08      14996.0   \n",
       "CASR            CASR        1.926148e-01      15482.0   \n",
       "CWC15          CWC15        6.812821e-01      11302.0   \n",
       "FAM110B      FAM110B        1.186656e-26      16007.0   \n",
       "DYNC1LI2    DYNC1LI2        4.908478e-02       5687.0   \n",
       "MUL1            MUL1        6.487886e-01       1893.0   \n",
       "\n",
       "           abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "VEZT                         0.564027               0.564027   \n",
       "THAP9                        0.399319               0.399319   \n",
       "TTC1                         0.434710               0.434710   \n",
       "LINC01730                    1.484307               1.484307   \n",
       "CIT                          2.899800               2.899800   \n",
       "CASR                         4.649033               4.649033   \n",
       "CWC15                        0.788082               0.788082   \n",
       "FAM110B                     21.693587             -21.693587   \n",
       "DYNC1LI2                     0.267402               0.267402   \n",
       "MUL1                         0.069030              -0.069030   \n",
       "\n",
       "           Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "VEZT                             0.045076            4234.0   \n",
       "THAP9                            0.784561             107.0   \n",
       "TTC1                             0.070044            1441.0   \n",
       "LINC01730                        0.757420             303.0   \n",
       "CIT                              0.254798            2050.0   \n",
       "CASR                             0.402032           10815.0   \n",
       "CWC15                            0.003572            3123.0   \n",
       "FAM110B                          0.081285            9364.0   \n",
       "DYNC1LI2                         0.328713             882.0   \n",
       "MUL1                             0.755263              56.0   \n",
       "\n",
       "           Percentile (simulated)  Percent DE (simulated)  \\\n",
       "VEZT                    23.842514                     0.0   \n",
       "THAP9                    0.597049                     0.0   \n",
       "TTC1                     8.110848                     0.0   \n",
       "LINC01730                1.701025                     0.0   \n",
       "CIT                     11.541061                     0.0   \n",
       "CASR                    60.910217                     0.0   \n",
       "CWC15                   17.584770                     0.0   \n",
       "FAM110B                 52.737411                     0.0   \n",
       "DYNC1LI2                 4.962262                     0.0   \n",
       "MUL1                     0.309789                     0.0   \n",
       "\n",
       "           Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "VEZT                                   0.236518                   0.000052   \n",
       "THAP9                                  0.027406                   0.000059   \n",
       "TTC1                                   0.115925                   0.000118   \n",
       "LINC01730                              0.050050                   0.000591   \n",
       "CIT                                    0.142667                   0.001702   \n",
       "CASR                                   0.903815                   0.002424   \n",
       "CWC15                                  0.187333                   0.000407   \n",
       "FAM110B                                0.614677                   0.015747   \n",
       "DYNC1LI2                               0.088109                   0.000135   \n",
       "MUL1                                   0.018416                   0.000040   \n",
       "\n",
       "           Number of experiments (simulated)      Z score  \n",
       "VEZT                                       2  6301.055203  \n",
       "THAP9                                      2  6257.662855  \n",
       "TTC1                                       2  2698.733956  \n",
       "LINC01730                                  2  2428.170510  \n",
       "CIT                                        2  1619.890114  \n",
       "CASR                                       2  1544.822271  \n",
       "CWC15                                      2  1474.512916  \n",
       "FAM110B                                    2  1338.586384  \n",
       "DYNC1LI2                                   2  1324.782792  \n",
       "MUL1                                       2  1266.050242  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                                 False\n",
       "Adj P-value (Real)                       True\n",
       "Rank (Real)                              True\n",
       "abs(log2FoldChange) (Real)               True\n",
       "log2FoldChange (Real)                    True\n",
       "Median adj p-value (simulated)          False\n",
       "Rank (simulated)                        False\n",
       "Percentile (simulated)                  False\n",
       "Percent DE (simulated)                  False\n",
       "Mean abs(log2FoldChange) (simulated)    False\n",
       "Std deviation (simulated)                True\n",
       "Number of experiments (simulated)       False\n",
       "Z score                                  True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>17140.0</td>\n",
       "      <td>96.535992</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.634457</td>\n",
       "      <td>3.969021</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB7</th>\n",
       "      <td>ABCB7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>14829.0</td>\n",
       "      <td>83.519207</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.869387</td>\n",
       "      <td>3.681591</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCG5</th>\n",
       "      <td>ABCG5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000847</td>\n",
       "      <td>16269.0</td>\n",
       "      <td>91.630055</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.849066</td>\n",
       "      <td>4.605803</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCG8</th>\n",
       "      <td>ABCG8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>15765.0</td>\n",
       "      <td>88.791258</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.507399</td>\n",
       "      <td>3.608295</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABRA</th>\n",
       "      <td>ABRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.002624</td>\n",
       "      <td>16148.0</td>\n",
       "      <td>90.948519</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.762359</td>\n",
       "      <td>4.433512</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOD1</th>\n",
       "      <td>ACOD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.598339</td>\n",
       "      <td>12512.0</td>\n",
       "      <td>70.468627</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.595422</td>\n",
       "      <td>2.149427</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACRV1</th>\n",
       "      <td>ACRV1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.342338</td>\n",
       "      <td>13511.0</td>\n",
       "      <td>76.095528</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.103757</td>\n",
       "      <td>2.895575</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSM2A</th>\n",
       "      <td>ACSM2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.306849</td>\n",
       "      <td>12223.0</td>\n",
       "      <td>68.840825</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.451264</td>\n",
       "      <td>0.736960</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTA1</th>\n",
       "      <td>ACTA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.046634</td>\n",
       "      <td>14490.0</td>\n",
       "      <td>81.609778</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.669649</td>\n",
       "      <td>2.018400</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTC1</th>\n",
       "      <td>ACTC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.133748</td>\n",
       "      <td>14478.0</td>\n",
       "      <td>81.542188</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.661343</td>\n",
       "      <td>1.433808</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL6B</th>\n",
       "      <td>ACTL6B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000361</td>\n",
       "      <td>17310.0</td>\n",
       "      <td>97.493523</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.932996</td>\n",
       "      <td>5.066015</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL7A</th>\n",
       "      <td>ACTL7A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.135578</td>\n",
       "      <td>16831.0</td>\n",
       "      <td>94.795539</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.328325</td>\n",
       "      <td>5.483952</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL7B</th>\n",
       "      <td>ACTL7B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.141864</td>\n",
       "      <td>17601.0</td>\n",
       "      <td>99.132590</td>\n",
       "      <td>0.5</td>\n",
       "      <td>9.194951</td>\n",
       "      <td>12.512321</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL8</th>\n",
       "      <td>ACTL8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>17545.0</td>\n",
       "      <td>98.817168</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.080625</td>\n",
       "      <td>5.528168</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL9</th>\n",
       "      <td>ACTL9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.247421</td>\n",
       "      <td>16169.0</td>\n",
       "      <td>91.066802</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.778266</td>\n",
       "      <td>5.012686</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTN3</th>\n",
       "      <td>ACTN3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.198931</td>\n",
       "      <td>17618.0</td>\n",
       "      <td>99.228343</td>\n",
       "      <td>0.5</td>\n",
       "      <td>9.316318</td>\n",
       "      <td>12.840983</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTRT2</th>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.081875</td>\n",
       "      <td>16708.0</td>\n",
       "      <td>94.102737</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.191690</td>\n",
       "      <td>5.013678</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAD1</th>\n",
       "      <td>ADAD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.370857</td>\n",
       "      <td>11947.0</td>\n",
       "      <td>67.286245</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.327659</td>\n",
       "      <td>0.208324</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM18</th>\n",
       "      <td>ADAM18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.014899</td>\n",
       "      <td>15336.0</td>\n",
       "      <td>86.374901</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.210049</td>\n",
       "      <td>2.685781</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM20</th>\n",
       "      <td>ADAM20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.191063</td>\n",
       "      <td>17573.0</td>\n",
       "      <td>98.974879</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8.825432</td>\n",
       "      <td>11.999660</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM3A</th>\n",
       "      <td>ADAM3A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011511</td>\n",
       "      <td>15392.0</td>\n",
       "      <td>86.690323</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.246001</td>\n",
       "      <td>2.996296</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM5</th>\n",
       "      <td>ADAM5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.402033</td>\n",
       "      <td>11762.0</td>\n",
       "      <td>66.244227</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.247283</td>\n",
       "      <td>0.483313</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM6</th>\n",
       "      <td>ADAM6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.016338</td>\n",
       "      <td>14859.0</td>\n",
       "      <td>83.688183</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.893960</td>\n",
       "      <td>2.845332</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM7</th>\n",
       "      <td>ADAM7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.125658</td>\n",
       "      <td>14741.0</td>\n",
       "      <td>83.023544</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.818069</td>\n",
       "      <td>1.390810</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADH4</th>\n",
       "      <td>ADH4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.321454</td>\n",
       "      <td>11955.0</td>\n",
       "      <td>67.331306</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.333347</td>\n",
       "      <td>0.415248</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADIG</th>\n",
       "      <td>ADIG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.009778</td>\n",
       "      <td>15211.0</td>\n",
       "      <td>85.670835</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.123612</td>\n",
       "      <td>3.134861</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADNP</th>\n",
       "      <td>ADNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.004877</td>\n",
       "      <td>14546.0</td>\n",
       "      <td>81.925200</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.692422</td>\n",
       "      <td>3.356292</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFM</th>\n",
       "      <td>AFM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.445926</td>\n",
       "      <td>11546.0</td>\n",
       "      <td>65.027599</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.159273</td>\n",
       "      <td>0.904797</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFP</th>\n",
       "      <td>AFP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.448140</td>\n",
       "      <td>9894.0</td>\n",
       "      <td>55.722654</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.696208</td>\n",
       "      <td>0.249925</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGXT</th>\n",
       "      <td>AGXT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.284974</td>\n",
       "      <td>12202.0</td>\n",
       "      <td>68.722541</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.441425</td>\n",
       "      <td>0.890576</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WNT8A</th>\n",
       "      <td>WNT8A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.199701</td>\n",
       "      <td>13707.0</td>\n",
       "      <td>77.199504</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.204867</td>\n",
       "      <td>2.801000</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XAGE3</th>\n",
       "      <td>XAGE3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.022467</td>\n",
       "      <td>15188.0</td>\n",
       "      <td>85.541286</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.105766</td>\n",
       "      <td>3.298491</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XIRP2</th>\n",
       "      <td>XIRP2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.116496</td>\n",
       "      <td>17586.0</td>\n",
       "      <td>99.048102</td>\n",
       "      <td>0.5</td>\n",
       "      <td>9.069822</td>\n",
       "      <td>12.071212</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC13</th>\n",
       "      <td>ZCCHC13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.067572</td>\n",
       "      <td>14798.0</td>\n",
       "      <td>83.344598</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.850223</td>\n",
       "      <td>2.044285</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZDHHC22</th>\n",
       "      <td>ZDHHC22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.121928</td>\n",
       "      <td>17000.0</td>\n",
       "      <td>95.747437</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.482879</td>\n",
       "      <td>6.004621</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFP42</th>\n",
       "      <td>ZFP42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.479017</td>\n",
       "      <td>12306.0</td>\n",
       "      <td>69.308325</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.489047</td>\n",
       "      <td>1.939967</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFY</th>\n",
       "      <td>ZFY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.028368</td>\n",
       "      <td>9197.0</td>\n",
       "      <td>51.796778</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.590660</td>\n",
       "      <td>0.519840</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZG16</th>\n",
       "      <td>ZG16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.307671</td>\n",
       "      <td>14901.0</td>\n",
       "      <td>83.924749</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.920253</td>\n",
       "      <td>1.652716</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZG16B</th>\n",
       "      <td>ZG16B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.129016</td>\n",
       "      <td>14408.0</td>\n",
       "      <td>81.147910</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.618530</td>\n",
       "      <td>1.589888</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZGLP1</th>\n",
       "      <td>ZGLP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.028794</td>\n",
       "      <td>15918.0</td>\n",
       "      <td>89.653036</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.603594</td>\n",
       "      <td>4.576573</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC2</th>\n",
       "      <td>ZIC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.132784</td>\n",
       "      <td>14847.0</td>\n",
       "      <td>83.620593</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.882521</td>\n",
       "      <td>3.827417</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC3</th>\n",
       "      <td>ZIC3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.177835</td>\n",
       "      <td>16603.0</td>\n",
       "      <td>93.511321</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.112323</td>\n",
       "      <td>5.510820</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIM3</th>\n",
       "      <td>ZIM3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.299596</td>\n",
       "      <td>12771.0</td>\n",
       "      <td>71.927453</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.724435</td>\n",
       "      <td>1.685936</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZKSCAN2</th>\n",
       "      <td>ZKSCAN2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.477177</td>\n",
       "      <td>17363.0</td>\n",
       "      <td>97.792047</td>\n",
       "      <td>0.5</td>\n",
       "      <td>5.053939</td>\n",
       "      <td>7.120499</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMAT4</th>\n",
       "      <td>ZMAT4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>16547.0</td>\n",
       "      <td>93.195900</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.060575</td>\n",
       "      <td>4.610013</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMYND11</th>\n",
       "      <td>ZMYND11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003862</td>\n",
       "      <td>16943.0</td>\n",
       "      <td>95.426383</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.416874</td>\n",
       "      <td>5.400532</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF280A</th>\n",
       "      <td>ZNF280A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001320</td>\n",
       "      <td>16569.0</td>\n",
       "      <td>93.319815</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.079823</td>\n",
       "      <td>4.089998</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF311</th>\n",
       "      <td>ZNF311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007175</td>\n",
       "      <td>17595.0</td>\n",
       "      <td>99.098795</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.155274</td>\n",
       "      <td>11.398898</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF32.AS3</th>\n",
       "      <td>ZNF32.AS3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.557754</td>\n",
       "      <td>8750.0</td>\n",
       "      <td>49.279036</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.535297</td>\n",
       "      <td>0.702440</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF488</th>\n",
       "      <td>ZNF488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017955</td>\n",
       "      <td>10387.0</td>\n",
       "      <td>58.499493</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.794153</td>\n",
       "      <td>0.655961</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF536</th>\n",
       "      <td>ZNF536</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.257627</td>\n",
       "      <td>17027.0</td>\n",
       "      <td>95.899516</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.498062</td>\n",
       "      <td>5.985577</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF556</th>\n",
       "      <td>ZNF556</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.009762</td>\n",
       "      <td>14805.0</td>\n",
       "      <td>83.384026</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.856776</td>\n",
       "      <td>3.487709</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF560</th>\n",
       "      <td>ZNF560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.110307</td>\n",
       "      <td>16310.0</td>\n",
       "      <td>91.860989</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.882456</td>\n",
       "      <td>5.176364</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF582</th>\n",
       "      <td>ZNF582</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.004739</td>\n",
       "      <td>17731.0</td>\n",
       "      <td>99.864819</td>\n",
       "      <td>0.5</td>\n",
       "      <td>10.138089</td>\n",
       "      <td>12.978043</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF669</th>\n",
       "      <td>ZNF669</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>17412.0</td>\n",
       "      <td>98.068041</td>\n",
       "      <td>0.5</td>\n",
       "      <td>5.191499</td>\n",
       "      <td>6.589888</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF880</th>\n",
       "      <td>ZNF880</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.153010</td>\n",
       "      <td>13546.0</td>\n",
       "      <td>76.292666</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.119962</td>\n",
       "      <td>2.339610</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF4</th>\n",
       "      <td>ZNRF4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.047745</td>\n",
       "      <td>16179.0</td>\n",
       "      <td>91.123127</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.786381</td>\n",
       "      <td>4.495543</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZP4</th>\n",
       "      <td>ZP4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.500060</td>\n",
       "      <td>10518.0</td>\n",
       "      <td>59.237355</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.824627</td>\n",
       "      <td>1.166198</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295559</td>\n",
       "      <td>15688.0</td>\n",
       "      <td>88.357553</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.453881</td>\n",
       "      <td>4.681960</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP2</th>\n",
       "      <td>ZPBP2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.156591</td>\n",
       "      <td>13991.0</td>\n",
       "      <td>78.799144</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.374039</td>\n",
       "      <td>1.390083</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1726 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "A1CF            A1CF                 NaN          NaN   \n",
       "ABCB7          ABCB7                 NaN          NaN   \n",
       "ABCG5          ABCG5                 NaN          NaN   \n",
       "ABCG8          ABCG8                 NaN          NaN   \n",
       "ABRA            ABRA                 NaN          NaN   \n",
       "ACOD1          ACOD1                 NaN          NaN   \n",
       "ACRV1          ACRV1                 NaN          NaN   \n",
       "ACSM2A        ACSM2A                 NaN          NaN   \n",
       "ACTA1          ACTA1                 NaN          NaN   \n",
       "ACTC1          ACTC1                 NaN          NaN   \n",
       "ACTL6B        ACTL6B                 NaN          NaN   \n",
       "ACTL7A        ACTL7A                 NaN          NaN   \n",
       "ACTL7B        ACTL7B                 NaN          NaN   \n",
       "ACTL8          ACTL8                 NaN          NaN   \n",
       "ACTL9          ACTL9                 NaN          NaN   \n",
       "ACTN3          ACTN3                 NaN          NaN   \n",
       "ACTRT2        ACTRT2                 NaN          NaN   \n",
       "ADAD1          ADAD1                 NaN          NaN   \n",
       "ADAM18        ADAM18                 NaN          NaN   \n",
       "ADAM20        ADAM20                 NaN          NaN   \n",
       "ADAM3A        ADAM3A                 NaN          NaN   \n",
       "ADAM5          ADAM5                 NaN          NaN   \n",
       "ADAM6          ADAM6                 NaN          NaN   \n",
       "ADAM7          ADAM7                 NaN          NaN   \n",
       "ADH4            ADH4                 NaN          NaN   \n",
       "ADIG            ADIG                 NaN          NaN   \n",
       "ADNP            ADNP                 NaN          NaN   \n",
       "AFM              AFM                 NaN          NaN   \n",
       "AFP              AFP                 NaN          NaN   \n",
       "AGXT            AGXT                 NaN          NaN   \n",
       "...              ...                 ...          ...   \n",
       "WNT8A          WNT8A                 NaN          NaN   \n",
       "XAGE3          XAGE3                 NaN          NaN   \n",
       "XIRP2          XIRP2                 NaN          NaN   \n",
       "ZCCHC13      ZCCHC13                 NaN          NaN   \n",
       "ZDHHC22      ZDHHC22                 NaN          NaN   \n",
       "ZFP42          ZFP42                 NaN          NaN   \n",
       "ZFY              ZFY                 NaN          NaN   \n",
       "ZG16            ZG16                 NaN          NaN   \n",
       "ZG16B          ZG16B                 NaN          NaN   \n",
       "ZGLP1          ZGLP1                 NaN          NaN   \n",
       "ZIC2            ZIC2                 NaN          NaN   \n",
       "ZIC3            ZIC3                 NaN          NaN   \n",
       "ZIM3            ZIM3                 NaN          NaN   \n",
       "ZKSCAN2      ZKSCAN2                 NaN          NaN   \n",
       "ZMAT4          ZMAT4                 NaN          NaN   \n",
       "ZMYND11      ZMYND11                 NaN          NaN   \n",
       "ZNF280A      ZNF280A                 NaN          NaN   \n",
       "ZNF311        ZNF311                 NaN          NaN   \n",
       "ZNF32.AS3  ZNF32.AS3                 NaN          NaN   \n",
       "ZNF488        ZNF488                 NaN          NaN   \n",
       "ZNF536        ZNF536                 NaN          NaN   \n",
       "ZNF556        ZNF556                 NaN          NaN   \n",
       "ZNF560        ZNF560                 NaN          NaN   \n",
       "ZNF582        ZNF582                 NaN          NaN   \n",
       "ZNF669        ZNF669                 NaN          NaN   \n",
       "ZNF880        ZNF880                 NaN          NaN   \n",
       "ZNRF4          ZNRF4                 NaN          NaN   \n",
       "ZP4              ZP4                 NaN          NaN   \n",
       "ZPBP            ZPBP                 NaN          NaN   \n",
       "ZPBP2          ZPBP2                 NaN          NaN   \n",
       "\n",
       "           abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "A1CF                              NaN                    NaN   \n",
       "ABCB7                             NaN                    NaN   \n",
       "ABCG5                             NaN                    NaN   \n",
       "ABCG8                             NaN                    NaN   \n",
       "ABRA                              NaN                    NaN   \n",
       "ACOD1                             NaN                    NaN   \n",
       "ACRV1                             NaN                    NaN   \n",
       "ACSM2A                            NaN                    NaN   \n",
       "ACTA1                             NaN                    NaN   \n",
       "ACTC1                             NaN                    NaN   \n",
       "ACTL6B                            NaN                    NaN   \n",
       "ACTL7A                            NaN                    NaN   \n",
       "ACTL7B                            NaN                    NaN   \n",
       "ACTL8                             NaN                    NaN   \n",
       "ACTL9                             NaN                    NaN   \n",
       "ACTN3                             NaN                    NaN   \n",
       "ACTRT2                            NaN                    NaN   \n",
       "ADAD1                             NaN                    NaN   \n",
       "ADAM18                            NaN                    NaN   \n",
       "ADAM20                            NaN                    NaN   \n",
       "ADAM3A                            NaN                    NaN   \n",
       "ADAM5                             NaN                    NaN   \n",
       "ADAM6                             NaN                    NaN   \n",
       "ADAM7                             NaN                    NaN   \n",
       "ADH4                              NaN                    NaN   \n",
       "ADIG                              NaN                    NaN   \n",
       "ADNP                              NaN                    NaN   \n",
       "AFM                               NaN                    NaN   \n",
       "AFP                               NaN                    NaN   \n",
       "AGXT                              NaN                    NaN   \n",
       "...                               ...                    ...   \n",
       "WNT8A                             NaN                    NaN   \n",
       "XAGE3                             NaN                    NaN   \n",
       "XIRP2                             NaN                    NaN   \n",
       "ZCCHC13                           NaN                    NaN   \n",
       "ZDHHC22                           NaN                    NaN   \n",
       "ZFP42                             NaN                    NaN   \n",
       "ZFY                               NaN                    NaN   \n",
       "ZG16                              NaN                    NaN   \n",
       "ZG16B                             NaN                    NaN   \n",
       "ZGLP1                             NaN                    NaN   \n",
       "ZIC2                              NaN                    NaN   \n",
       "ZIC3                              NaN                    NaN   \n",
       "ZIM3                              NaN                    NaN   \n",
       "ZKSCAN2                           NaN                    NaN   \n",
       "ZMAT4                             NaN                    NaN   \n",
       "ZMYND11                           NaN                    NaN   \n",
       "ZNF280A                           NaN                    NaN   \n",
       "ZNF311                            NaN                    NaN   \n",
       "ZNF32.AS3                         NaN                    NaN   \n",
       "ZNF488                            NaN                    NaN   \n",
       "ZNF536                            NaN                    NaN   \n",
       "ZNF556                            NaN                    NaN   \n",
       "ZNF560                            NaN                    NaN   \n",
       "ZNF582                            NaN                    NaN   \n",
       "ZNF669                            NaN                    NaN   \n",
       "ZNF880                            NaN                    NaN   \n",
       "ZNRF4                             NaN                    NaN   \n",
       "ZP4                               NaN                    NaN   \n",
       "ZPBP                              NaN                    NaN   \n",
       "ZPBP2                             NaN                    NaN   \n",
       "\n",
       "           Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "A1CF                             0.000018           17140.0   \n",
       "ABCB7                            0.000667           14829.0   \n",
       "ABCG5                            0.000847           16269.0   \n",
       "ABCG8                            0.000624           15765.0   \n",
       "ABRA                             0.002624           16148.0   \n",
       "ACOD1                            0.598339           12512.0   \n",
       "ACRV1                            0.342338           13511.0   \n",
       "ACSM2A                           0.306849           12223.0   \n",
       "ACTA1                            0.046634           14490.0   \n",
       "ACTC1                            0.133748           14478.0   \n",
       "ACTL6B                           0.000361           17310.0   \n",
       "ACTL7A                           0.135578           16831.0   \n",
       "ACTL7B                           0.141864           17601.0   \n",
       "ACTL8                            0.000001           17545.0   \n",
       "ACTL9                            0.247421           16169.0   \n",
       "ACTN3                            0.198931           17618.0   \n",
       "ACTRT2                           0.081875           16708.0   \n",
       "ADAD1                            0.370857           11947.0   \n",
       "ADAM18                           0.014899           15336.0   \n",
       "ADAM20                           0.191063           17573.0   \n",
       "ADAM3A                           0.011511           15392.0   \n",
       "ADAM5                            0.402033           11762.0   \n",
       "ADAM6                            0.016338           14859.0   \n",
       "ADAM7                            0.125658           14741.0   \n",
       "ADH4                             0.321454           11955.0   \n",
       "ADIG                             0.009778           15211.0   \n",
       "ADNP                             0.004877           14546.0   \n",
       "AFM                              0.445926           11546.0   \n",
       "AFP                              0.448140            9894.0   \n",
       "AGXT                             0.284974           12202.0   \n",
       "...                                   ...               ...   \n",
       "WNT8A                            0.199701           13707.0   \n",
       "XAGE3                            0.022467           15188.0   \n",
       "XIRP2                            0.116496           17586.0   \n",
       "ZCCHC13                          0.067572           14798.0   \n",
       "ZDHHC22                          0.121928           17000.0   \n",
       "ZFP42                            0.479017           12306.0   \n",
       "ZFY                              0.028368            9197.0   \n",
       "ZG16                             0.307671           14901.0   \n",
       "ZG16B                            0.129016           14408.0   \n",
       "ZGLP1                            0.028794           15918.0   \n",
       "ZIC2                             0.132784           14847.0   \n",
       "ZIC3                             0.177835           16603.0   \n",
       "ZIM3                             0.299596           12771.0   \n",
       "ZKSCAN2                          0.477177           17363.0   \n",
       "ZMAT4                            0.000329           16547.0   \n",
       "ZMYND11                          0.003862           16943.0   \n",
       "ZNF280A                          0.001320           16569.0   \n",
       "ZNF311                           0.007175           17595.0   \n",
       "ZNF32.AS3                        0.557754            8750.0   \n",
       "ZNF488                           0.017955           10387.0   \n",
       "ZNF536                           0.257627           17027.0   \n",
       "ZNF556                           0.009762           14805.0   \n",
       "ZNF560                           0.110307           16310.0   \n",
       "ZNF582                           0.004739           17731.0   \n",
       "ZNF669                           0.000128           17412.0   \n",
       "ZNF880                           0.153010           13546.0   \n",
       "ZNRF4                            0.047745           16179.0   \n",
       "ZP4                              0.500060           10518.0   \n",
       "ZPBP                             0.295559           15688.0   \n",
       "ZPBP2                            0.156591           13991.0   \n",
       "\n",
       "           Percentile (simulated)  Percent DE (simulated)  \\\n",
       "A1CF                    96.535992                     1.0   \n",
       "ABCB7                   83.519207                     0.5   \n",
       "ABCG5                   91.630055                     0.5   \n",
       "ABCG8                   88.791258                     0.5   \n",
       "ABRA                    90.948519                     0.5   \n",
       "ACOD1                   70.468627                     0.0   \n",
       "ACRV1                   76.095528                     0.5   \n",
       "ACSM2A                  68.840825                     0.0   \n",
       "ACTA1                   81.609778                     0.5   \n",
       "ACTC1                   81.542188                     0.5   \n",
       "ACTL6B                  97.493523                     1.0   \n",
       "ACTL7A                  94.795539                     0.5   \n",
       "ACTL7B                  99.132590                     0.5   \n",
       "ACTL8                   98.817168                     1.0   \n",
       "ACTL9                   91.066802                     0.5   \n",
       "ACTN3                   99.228343                     0.5   \n",
       "ACTRT2                  94.102737                     0.5   \n",
       "ADAD1                   67.286245                     0.5   \n",
       "ADAM18                  86.374901                     1.0   \n",
       "ADAM20                  98.974879                     0.5   \n",
       "ADAM3A                  86.690323                     1.0   \n",
       "ADAM5                   66.244227                     0.5   \n",
       "ADAM6                   83.688183                     0.5   \n",
       "ADAM7                   83.023544                     0.5   \n",
       "ADH4                    67.331306                     0.5   \n",
       "ADIG                    85.670835                     0.5   \n",
       "ADNP                    81.925200                     0.5   \n",
       "AFM                     65.027599                     0.5   \n",
       "AFP                     55.722654                     0.0   \n",
       "AGXT                    68.722541                     0.0   \n",
       "...                           ...                     ...   \n",
       "WNT8A                   77.199504                     0.5   \n",
       "XAGE3                   85.541286                     0.5   \n",
       "XIRP2                   99.048102                     0.5   \n",
       "ZCCHC13                 83.344598                     0.5   \n",
       "ZDHHC22                 95.747437                     0.5   \n",
       "ZFP42                   69.308325                     0.0   \n",
       "ZFY                     51.796778                     0.0   \n",
       "ZG16                    83.924749                     0.5   \n",
       "ZG16B                   81.147910                     0.5   \n",
       "ZGLP1                   89.653036                     0.5   \n",
       "ZIC2                    83.620593                     0.5   \n",
       "ZIC3                    93.511321                     0.5   \n",
       "ZIM3                    71.927453                     0.0   \n",
       "ZKSCAN2                 97.792047                     0.5   \n",
       "ZMAT4                   93.195900                     0.5   \n",
       "ZMYND11                 95.426383                     0.5   \n",
       "ZNF280A                 93.319815                     1.0   \n",
       "ZNF311                  99.098795                     1.0   \n",
       "ZNF32.AS3               49.279036                     0.0   \n",
       "ZNF488                  58.499493                     0.5   \n",
       "ZNF536                  95.899516                     0.5   \n",
       "ZNF556                  83.384026                     0.5   \n",
       "ZNF560                  91.860989                     0.5   \n",
       "ZNF582                  99.864819                     0.5   \n",
       "ZNF669                  98.068041                     0.5   \n",
       "ZNF880                  76.292666                     0.0   \n",
       "ZNRF4                   91.123127                     0.5   \n",
       "ZP4                     59.237355                     0.5   \n",
       "ZPBP                    88.357553                     0.5   \n",
       "ZPBP2                   78.799144                     0.5   \n",
       "\n",
       "           Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "A1CF                                   4.634457                   3.969021   \n",
       "ABCB7                                  2.869387                   3.681591   \n",
       "ABCG5                                  3.849066                   4.605803   \n",
       "ABCG8                                  3.507399                   3.608295   \n",
       "ABRA                                   3.762359                   4.433512   \n",
       "ACOD1                                  1.595422                   2.149427   \n",
       "ACRV1                                  2.103757                   2.895575   \n",
       "ACSM2A                                 1.451264                   0.736960   \n",
       "ACTA1                                  2.669649                   2.018400   \n",
       "ACTC1                                  2.661343                   1.433808   \n",
       "ACTL6B                                 4.932996                   5.066015   \n",
       "ACTL7A                                 4.328325                   5.483952   \n",
       "ACTL7B                                 9.194951                  12.512321   \n",
       "ACTL8                                  6.080625                   5.528168   \n",
       "ACTL9                                  3.778266                   5.012686   \n",
       "ACTN3                                  9.316318                  12.840983   \n",
       "ACTRT2                                 4.191690                   5.013678   \n",
       "ADAD1                                  1.327659                   0.208324   \n",
       "ADAM18                                 3.210049                   2.685781   \n",
       "ADAM20                                 8.825432                  11.999660   \n",
       "ADAM3A                                 3.246001                   2.996296   \n",
       "ADAM5                                  1.247283                   0.483313   \n",
       "ADAM6                                  2.893960                   2.845332   \n",
       "ADAM7                                  2.818069                   1.390810   \n",
       "ADH4                                   1.333347                   0.415248   \n",
       "ADIG                                   3.123612                   3.134861   \n",
       "ADNP                                   2.692422                   3.356292   \n",
       "AFM                                    1.159273                   0.904797   \n",
       "AFP                                    0.696208                   0.249925   \n",
       "AGXT                                   1.441425                   0.890576   \n",
       "...                                         ...                        ...   \n",
       "WNT8A                                  2.204867                   2.801000   \n",
       "XAGE3                                  3.105766                   3.298491   \n",
       "XIRP2                                  9.069822                  12.071212   \n",
       "ZCCHC13                                2.850223                   2.044285   \n",
       "ZDHHC22                                4.482879                   6.004621   \n",
       "ZFP42                                  1.489047                   1.939967   \n",
       "ZFY                                    0.590660                   0.519840   \n",
       "ZG16                                   2.920253                   1.652716   \n",
       "ZG16B                                  2.618530                   1.589888   \n",
       "ZGLP1                                  3.603594                   4.576573   \n",
       "ZIC2                                   2.882521                   3.827417   \n",
       "ZIC3                                   4.112323                   5.510820   \n",
       "ZIM3                                   1.724435                   1.685936   \n",
       "ZKSCAN2                                5.053939                   7.120499   \n",
       "ZMAT4                                  4.060575                   4.610013   \n",
       "ZMYND11                                4.416874                   5.400532   \n",
       "ZNF280A                                4.079823                   4.089998   \n",
       "ZNF311                                 9.155274                  11.398898   \n",
       "ZNF32.AS3                              0.535297                   0.702440   \n",
       "ZNF488                                 0.794153                   0.655961   \n",
       "ZNF536                                 4.498062                   5.985577   \n",
       "ZNF556                                 2.856776                   3.487709   \n",
       "ZNF560                                 3.882456                   5.176364   \n",
       "ZNF582                                10.138089                  12.978043   \n",
       "ZNF669                                 5.191499                   6.589888   \n",
       "ZNF880                                 2.119962                   2.339610   \n",
       "ZNRF4                                  3.786381                   4.495543   \n",
       "ZP4                                    0.824627                   1.166198   \n",
       "ZPBP                                   3.453881                   4.681960   \n",
       "ZPBP2                                  2.374039                   1.390083   \n",
       "\n",
       "           Number of experiments (simulated)  Z score  \n",
       "A1CF                                       2      NaN  \n",
       "ABCB7                                      2      NaN  \n",
       "ABCG5                                      2      NaN  \n",
       "ABCG8                                      2      NaN  \n",
       "ABRA                                       2      NaN  \n",
       "ACOD1                                      2      NaN  \n",
       "ACRV1                                      2      NaN  \n",
       "ACSM2A                                     2      NaN  \n",
       "ACTA1                                      2      NaN  \n",
       "ACTC1                                      2      NaN  \n",
       "ACTL6B                                     2      NaN  \n",
       "ACTL7A                                     2      NaN  \n",
       "ACTL7B                                     2      NaN  \n",
       "ACTL8                                      2      NaN  \n",
       "ACTL9                                      2      NaN  \n",
       "ACTN3                                      2      NaN  \n",
       "ACTRT2                                     2      NaN  \n",
       "ADAD1                                      2      NaN  \n",
       "ADAM18                                     2      NaN  \n",
       "ADAM20                                     2      NaN  \n",
       "ADAM3A                                     2      NaN  \n",
       "ADAM5                                      2      NaN  \n",
       "ADAM6                                      2      NaN  \n",
       "ADAM7                                      2      NaN  \n",
       "ADH4                                       2      NaN  \n",
       "ADIG                                       2      NaN  \n",
       "ADNP                                       2      NaN  \n",
       "AFM                                        2      NaN  \n",
       "AFP                                        2      NaN  \n",
       "AGXT                                       2      NaN  \n",
       "...                                      ...      ...  \n",
       "WNT8A                                      2      NaN  \n",
       "XAGE3                                      2      NaN  \n",
       "XIRP2                                      2      NaN  \n",
       "ZCCHC13                                    2      NaN  \n",
       "ZDHHC22                                    2      NaN  \n",
       "ZFP42                                      2      NaN  \n",
       "ZFY                                        2      NaN  \n",
       "ZG16                                       2      NaN  \n",
       "ZG16B                                      2      NaN  \n",
       "ZGLP1                                      2      NaN  \n",
       "ZIC2                                       2      NaN  \n",
       "ZIC3                                       2      NaN  \n",
       "ZIM3                                       2      NaN  \n",
       "ZKSCAN2                                    2      NaN  \n",
       "ZMAT4                                      2      NaN  \n",
       "ZMYND11                                    2      NaN  \n",
       "ZNF280A                                    2      NaN  \n",
       "ZNF311                                     2      NaN  \n",
       "ZNF32.AS3                                  2      NaN  \n",
       "ZNF488                                     2      NaN  \n",
       "ZNF536                                     2      NaN  \n",
       "ZNF556                                     2      NaN  \n",
       "ZNF560                                     2      NaN  \n",
       "ZNF582                                     2      NaN  \n",
       "ZNF669                                     2      NaN  \n",
       "ZNF880                                     2      NaN  \n",
       "ZNRF4                                      2      NaN  \n",
       "ZP4                                        2      NaN  \n",
       "ZPBP                                       2      NaN  \n",
       "ZPBP2                                      2      NaN  \n",
       "\n",
       "[1726 rows x 13 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks[summary_gene_ranks.isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "summary_gene_ranks.to_csv(output_filename, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:sophie] *",
   "language": "python",
   "name": "conda-env-sophie-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
